slick/src/main/scala/slick/util/AsyncExecutor.scala

```diff
@@ -78,17 +97,68 @@ object AsyncExecutor extends Logging {
     * @param maxThreads The maximum number of threads in the pool.
     * @param queueSize The size of the job queue, 0 for direct hand-off or -1 for unlimited size.
     * @param maxConnections The maximum number of configured connections for the connection pool.
-    *                       The underlying ThreadPoolExecutor will not pick up any more work when all connections are in use.
-    *                       It will resume as soon as a connection is released again to the pool
+    *                       The underlying ThreadPoolExecutor will not pick up any more work when all connections are in
+    *                       use. It will resume as soon as a connection is released again to the pool
     * @param keepAliveTime when the number of threads is greater than
     *        the core, this is the maximum time that excess idle threads
     *        will wait for new tasks before terminating.
     * @param registerMbeans If set to true, register an MXBean that provides insight into the current
     *        queue and thread pool workload. */
-  def apply(name: String, minThreads: Int, maxThreads: Int, queueSize: Int, maxConnections: Int, keepAliveTime: Duration,
+  def apply(name: String,
+            minThreads: Int,
+            maxThreads: Int,
+            queueSize: Int,
+            maxConnections: Int,
+            keepAliveTime: Duration,
             registerMbeans: Boolean): AsyncExecutor = new AsyncExecutor {
-    
-    @volatile private[this] lazy val mbeanName = new ObjectName(s"slick:type=AsyncExecutor,name=$name");
+
+    class ConfiguredSlickThreadPoolExecutor(queue: BlockingQueue[Runnable])
+      extends ThreadPoolExecutor(
+        minThreads,
+        maxThreads,
+        keepAliveTime.toMillis,
+        TimeUnit.MILLISECONDS,
+        queue,
+        new DaemonThreadFactory(name + "-")
+      )
+
+    class ManagedConfiguredSlickThreadPoolExecutor(queue: ManagedArrayBlockingQueue)
+      extends ConfiguredSlickThreadPoolExecutor(queue.asInstanceOf[BlockingQueue[Runnable]]) {
+      /** If the runnable/task is a low/medium priority item, we increase the items in use count, because first
+       * thing it will do is open a Jdbc connection from the pool.
+       */
+      override def beforeExecute(t: Thread, r: Runnable): Unit = {
+        r match {
+          case pr: PrioritizedRunnable if pr.priority != WithConnection =>
+            try queue.increaseInUseCount(pr)
+            catch {
+              case _: ManagedArrayBlockingQueue.CountLimitReachedException =>
+                logger.warn("Could not increase in-use count, because the limit has been reached. Will resubmit...")
+                executor.execute(r)
+            }
+          case _                                                        =>
+        }
+        super.beforeExecute(t, r)
+      }
```

Comment:
When we re-submit a task to the thread pool, we would also need to cancel its current execution

Reply:
Good point ;)

Reply:
I guess it should rethrow then? Not sure how else to prevent its execution.

Reply:
I don't understand this problem well enough... maybe we should have a chat on discord or something

Reply:
Absolutely, can you ping me there?